src/wasm32.rs

```diff
@@ -0,0 +1,563 @@
+        [i16; 4],
+        [u32; 2],
+        [i32; 2],
+        [f32; 2],
+        [u64; 1],
+        [i64; 1],
+        [f64; 1],
+        f64,
+        u64,
+        i64,
+    }
+}
+
+impl_N_bits_traits! {
+    impl Is16BytesUnaligned [16] for {
+        [u8; 16],
+        [i8; 16],
+        [u16; 8],
+        [i16; 8],
+        [u32; 4],
+        [i32; 4],
+        [f32; 4],
+        [u64; 2],
+        [i64; 2],
+        [f64; 2],
+    }
+}
+
+impl<T, const N: usize> private::Sealed for [core::cell::Cell<T>; N] where [T; N]: private::Sealed {}
+impl<T> private::Sealed for core::cell::Cell<T> where T: private::Sealed {}
+
+// Mark cell types as valid for their respective sizes. The wasm32 platform has many more load
+// operations than stores, so in contrast to other platforms, for brevity we implement the size
+// traits for cells instead of bounding the inner / array types in a separate trait. This makes
+// Cells redundant for stores (due to Cell::get_mut) but makes the list of exposed loads more
+// terse.
+//
+// As we mark the cell type, and arrays of such cells, the specific underlying types must be chosen
+// to avoid overlapping impls. This is a dance around the type checker as the many recursive impls
+// have potential of infinitely recursing until the next generation. Hence we make sure to always
+// reduce the matched type: both [Cell<T>; N] and Cell<T> take out a cell without introducing more
+// types to the obligations. E.g. Cell<[T; N]> for [T: N]  would not work.
+
+impl<T, const N: usize> Is1ByteUnaligned for [core::cell::Cell<T>; N] where [T; N]: Is1ByteUnaligned {}
+
+impl<T> Is1ByteUnaligned for core::cell::Cell<T> where T: Is1ByteUnaligned {}
+
+impl<T, const N: usize> Is2BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is2BytesUnaligned
+{
+}
+
+impl<T> Is2BytesUnaligned for core::cell::Cell<T> where T: Is2BytesUnaligned {}
+
+impl<T, const N: usize> Is4BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is4BytesUnaligned
+{
+}
+
+impl<T> Is4BytesUnaligned for core::cell::Cell<T> where T: Is4BytesUnaligned {}
+
+impl<T, const N: usize> Is8BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is8BytesUnaligned
+{
+}
+
+impl<T> Is8BytesUnaligned for core::cell::Cell<T> where T: Is8BytesUnaligned {}
+
+impl<T, const N: usize> Is16BytesUnaligned for [core::cell::Cell<T>; N] where
+    [T; N]: Is16BytesUnaligned
+{
+}
+
+impl<T> Is16BytesUnaligned for core::cell::Cell<T> where T: Is16BytesUnaligned {}
+
+#[cfg(test)]
+mod tests {
+    use core::arch::wasm32::{self as arch, v128};
+
+    fn assert_v128_bytes<const N: usize>(val: v128, data: &[[u8; N]]) {
+        assert_eq!(
+            unsafe { core::mem::transmute::<_, [u8; 16]>(val) },
+            unsafe { core::slice::from_raw_parts(data.as_ptr() as *const u8, data.len() * N) }
+        );
+
+        // assert_eq!(arch::to_bytes(val), <[u8; N]>::flat(data))
+    }
+
+    #[test]
+    #[cfg_attr(not(target_feature = "simd128"), ignore)]
```

Comment:
`cfg_attr` because it's not a baseline included feature, right?

Unlike the other arches

Reply:
Yes, for neither of wasm32-unknown and wasm32-wasip1, wasm32-wasip2. However, strangely, wasmtime does not seem to care. Also it is odd that we can execute functions with that additional target feature without an `unsafe` block. But miri loudly complains and does not support that target feature at all.

Reply:
According to: https://github.com/rust-lang/rust/issues/74372#issuecomment-828696632

Actually linking a function that made use of `target_feature(enable = )` requires that feature to be enabled globally, due to wasm specifying that validation happens _always. Hence, the only way to reach a function and execute is if the feature was enabled by the runtime, consequently the safety precondition is already fulfilled by `_main` being executed at all.

Seems sketch tbh but alright.